### PR TITLE
AP-253 Ensure back links on provider journey work as expected

### PR DESCRIPTION
**Pull request by stephenrichards**
This PR is to ensure all the back links go to the appropriate page
depending on what previous answers were given.

## What

[Link to story](https://dsdmoj.atlassian.net/browse/AP-253)

Updated steps.rb, and where necessary the logic in the controllers.
Added back link tests in all the request specs

## Checklist

Before you ask people to review this PR:

- [x] Tests and rubocop should be passing: `bundle exec rake`
- [x] Github should not be reporting conflicts; you should have recently run `git rebase master`.
- [x] There should be no unneccessary whitespace changes. These make diffs harder to read and conflicts more likely. 
- [x] The PR description should say what you changed and why, with a link to the JIRA story.
- [x] You should have looked at the diff against master and ensured that nothing unexpected is included in your changes.
- [x] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: app/views/providers/check_provider_answers/index.html.erb [hunk 1]
```diff
@@ -2,10 +2,10 @@
   content_for(:navigation) do
     link_to(
       t('generic.back'),
-      reset_providers_legal_aid_application_check_provider_answers_path,
+      @back_step_url,
```
Comment (Kalagan) [inline app/views/providers/check_provider_answers/index.html.erb]: If you remove the `reset...`, the application is stuck in "check_answers" state.
When on "Check answers" page, try clicking "back" twice, you should go back 2 steps (to `other_asserts` or `savings_and_investments`).
But with this change you go back to the "Check answers" page.

File: spec/requests/providers/savings_and_investments_spec.rb [hunk 2]
```diff
@@ -27,6 +27,29 @@
...
+          let(:application) { create :legal_aid_application, :without_own_home }
+          it 'points to the own  home page' do
+            expect(response.body).to have_back_link(providers_legal_aid_application_own_home_path(application))
+          end
+        end
+
+        context 'applicant owns home with shared ownership' do
+          let(:application) { create :legal_aid_application, :with_own_home_mortgaged, :with_home_shared_with_partner }
+          it 'points to the own  home page' do
+            expect(response.body).to have_back_link(providers_legal_aid_application_percentage_home_path(application))
```
Comment (Kalagan) [inline spec/requests/providers/savings_and_investments_spec.rb]: I think the description of this test is not correct.
Should it be `it 'points to the percentage_home page' do`?

File: spec/requests/providers/savings_and_investments_spec.rb [hunk 3]
```diff
@@ -27,6 +27,29 @@
...
+          end
+        end
+
+        context 'applicant owns home sole ownership' do
+          let(:application) { create :legal_aid_application, :with_own_home_mortgaged, :with_home_sole_owner }
+          it 'points to the own  home page' do
+            expect(response.body).to have_back_link(providers_legal_aid_application_shared_ownership_path(application))
```
Comment (Kalagan) [inline spec/requests/providers/savings_and_investments_spec.rb]: I think the description of this test is not correct.
Should it be `it 'points to the shared_ownership page' do`?

File: spec/requests/providers/shared_ownerships_spec.rb [hunk 4]
```diff
@@ -26,6 +26,22 @@
         expect(unescaped_response_body).to match('Does your client own their home with anyone else?')
         expect(unescaped_response_body).to match(LegalAidApplication::SHARED_OWNERSHIP_YES_REASONS.first)
       end
+
+      describe 'back link' do
+        context 'applicant owns with mortgage' do
+          it 'points to property_value page' do
+            expect(response.body).to have_back_link(providers_legal_aid_application_outstanding_mortgage_path(legal_aid_application))
```
Comment (Kalagan) [inline spec/requests/providers/shared_ownerships_spec.rb]: I think the description of this test is not correct.
Should it be it 'points to the outstanding_mortgage page' do?

File: spec/requests/providers/shared_ownerships_spec.rb [hunk 5]
```diff
@@ -26,6 +26,22 @@
...
+          end
+        end
+
+        context 'applicant owns home outright' do
+          let(:legal_aid_application) { create :legal_aid_application, :with_own_home_owned_outright }
+          it 'points to the outstanding mortgage page' do
+            allow(legal_aid_application).to receive(:own_home_mortgage?).and_return(true)
+            expect(response.body).to have_back_link(providers_legal_aid_application_property_value_path(legal_aid_application))
```
Comment (Kalagan) [inline spec/requests/providers/shared_ownerships_spec.rb]: I think the description of this test is not correct.
Should it be it 'points to the property_value page' do?

File: spec/support/back_link_matcher.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+RSpec::Matchers.define :have_back_link do |expected|
```
Comment (Kalagan) [inline spec/support/back_link_matcher.rb]: nice! 👍

File: spec/requests/providers/shared_ownerships_spec.rb (hunk 5 above)
Comment (stephenrichards) [reply spec/requests/providers/shared_ownerships_spec.rb]: yep

File: spec/requests/providers/shared_ownerships_spec.rb (hunk 4 above)
Comment (stephenrichards) [reply spec/requests/providers/shared_ownerships_spec.rb]: yep

File: spec/requests/providers/savings_and_investments_spec.rb (hunk 3 above)
Comment (stephenrichards) [reply spec/requests/providers/savings_and_investments_spec.rb]: yep

File: app/views/providers/check_provider_answers/index.html.erb (hunk 1 above)
Comment (stephenrichards) [reply app/views/providers/check_provider_answers/index.html.erb]: yep

File: spec/requests/providers/savings_and_investments_spec.rb (hunk 2 above)
Comment (stephenrichards) [reply spec/requests/providers/savings_and_investments_spec.rb]: yep

File: app/views/providers/check_provider_answers/index.html.erb [hunk 7]
```diff
@@ -84,5 +84,5 @@
                method: :patch,
                show_draft: true
              } %>
-  </div>
+</div>
 </div>
```
Comment (Kalagan) [inline app/views/providers/check_provider_answers/index.html.erb]: did you remove the indentation of this `</div>` by mistake?

**Pull request merged**